cmd/stargazer/main.go

```diff
@@ -0,0 +1,117 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The main package for the Stargazer application.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"github.com/nimbess/stargazer/pkg/config"
+	"github.com/nimbess/stargazer/pkg/controller"
+	"github.com/nimbess/stargazer/pkg/controllers/node"
+	log "github.com/sirupsen/logrus"
+	"os"
+	"strings"
+)
+
+// VERSION is updated during the build process using git
+var VERSION = "0.0.1"
+var version = false
+var cfgName = "stargazer"
+var cfgPath = "."
+var verbose = false
+
+func init() {
+	log.SetReportCaller(true)
+	flag.BoolVar(&version, "v", version, "Display version")
+	flag.StringVar(&cfgPath, "config-path", cfgPath, "Stargazer config file path")
+	flag.StringVar(&cfgName, "config-name", cfgName, "Stargazer config file name")
+	flag.BoolVar(&verbose, "verbose", verbose, "Enable debug logging")
+}
+
+func main() {
+	flag.Parse()
+	if version {
+		fmt.Println(VERSION)
+		os.Exit(0)
+	}
+
+	//cfg := new(config.Config)
+	cfg := config.NewConfig()
+	if err := cfg.Parse(cfgPath, cfgName); err != nil {
+		log.WithField("cfgName", cfgName).WithField("cfgPath", cfgPath).
+			WithError(err).Warn("Failed to parse config, using defaults")
+	}
+
+	log.WithField("config", fmt.Sprintf("%+v",*cfg)).Info("Configuration loaded")
+
+	logLevel, err := log.ParseLevel(cfg.LogLevel)
+	if err != nil {
+		logLevel = log.InfoLevel
+	}
+	log.SetLevel(logLevel)
+
+	stop := make(chan struct{})
+	defer close(stop)
+	ctx := context.Background()
+	// TODO: ensure connection to etcd is available
+
+	controllerCtrl := &controllerControl{
+		ctx:              ctx,
+		controllerStates: make(map[string]*controllerState),
+		config:           cfg,
+		stop:             stop,
+	}
+
+	// Create and start each requested controller
+	for _, controllerType := range strings.Split(cfg.Controllers, ",") {
+		switch controllerType {
+		case "node":
+			nodeController := node.NewNodeController(ctx, cfg)
+			controllerCtrl.controllerStates["Node"] = &controllerState {
+				controller: nodeController,
+				threadiness: cfg.NodeWorkers,
+			}
+		default:
+			log.WithField("controller", controllerType).Info("Invalid controller")
+		}
+	}
+
+	controllerCtrl.RunControllers()
+}
+
+// Object for keeping track of controller states and statuses.
+type controllerControl struct {
+	ctx              context.Context
+	controllerStates map[string]*controllerState
+	config           *config.Config
+	stop             chan struct{}
+}
+
+// Runs all the controllers and blocks indefinitely.
+func (cc *controllerControl) RunControllers() {
+	for controllerType, cs := range cc.controllerStates {
+		log.WithField("ControllerType", controllerType).Info("Starting controller")
+		go cs.controller.Run(cs.threadiness, cc.stop)
+	}
+	select {}
+}
+
+// Object for keeping track of Controller information.
+type controllerState struct {
```

Comment:
This name seems like it doesn't really fit. This doesn't maintain a state variable or anything, it just tracks a controller and how many threads it supports?

Reply:
The Controller is just an interface with a Run function. This is the controller itself, so it sounds like that is state since it's all I have. I can't think of a better name but am not tied to it either.

Reply:
Any struct which implements Run will be a Controller. You already do that with Node Controller. This is just a struct holding a controller and threadiness. If your goal is for this object to be a controller, then I think it would be better to just include your threadiness field inside of the controller struct itself, and then you don't need this struct at all.

Reply:
Renamed ControllerState to ControllerInfo since it holds the controller along with the request number of workers. This ControllerInfo is used with the ControllerControl map of controllerType,Info to start each controller type requested via the configuration.